bin/jsoo_main.ml

```diff
@@ -35,10 +35,34 @@ module Melange_ast = struct
     Melange_compiler_libs.Parsetree.structure = "%identity"
 end
 
+let last_warning : Location.report option ref = ref None
+
+(* We need to overload the original warning printer to capture the warnings
+   and not let them go through default printer (which will end up in browser console) *)
+let playground_warning_reporter (loc : Location.t) w : Location.report option =
+  let mk is_error id =
+    if is_error then Location.Report_warning_as_error id else Report_warning id
+  in
+  match Warnings.report w with
+  | `Inactive -> None
+  | `Active { Warnings.id; message; is_error; sub_locs } ->
+      let msg_of_str str ppf = Format.pp_print_string ppf str in
+      let kind = mk is_error id in
+      let main = { Location.loc; txt = msg_of_str message } in
+      let sub =
+        List.map
+          (fun (loc, sub_message) ->
+            { Location.loc; txt = msg_of_str sub_message })
+          sub_locs
+      in
+      last_warning := Some { Location.kind; main; sub };
```

Comment:
- can we collect warnings in a list or a queue instead? there might be more than one
- I couldn't find where `last_warning` gets reset? if we compile again, won't we show stale warnings?

Reply:
> can we collect warnings in a list or a queue instead? there might be more than one

I will check how to do this (will add a test to exercize it too).

> I couldn't find where `last_warning` gets reset? if we compile again, won't we show stale warnings?

I was doing it here:

https://github.com/melange-re/melange/blob/5d2731d90dd8baf0500bb1bfc69ea86dccc140c1/bin/jsoo_main.ml#L146

but this is probably not right. E.g. if we got a regular error + a warning, the value of `last_warning` would not be reset, as you mentioned. Will think about this as well.